common/src/main/java/com/microsoft/identity/common/crypto/AndroidBrokerStorageEncryptionManager.java

```diff
@@ -0,0 +1,168 @@
+import com.microsoft.identity.common.adal.internal.cache.StorageHelper;
+import com.microsoft.identity.common.java.crypto.StorageEncryptionManager;
+import com.microsoft.identity.common.java.crypto.key.AES256KeyLoader;
+import com.microsoft.identity.common.java.exception.ClientException;
+import com.microsoft.identity.common.java.telemetry.ITelemetryCallback;
+import com.microsoft.identity.common.logging.Logger;
+
+import java.util.ArrayList;
+import java.util.List;
+
+import lombok.NonNull;
+
+import static com.microsoft.identity.common.adal.internal.AuthenticationConstants.Broker.AZURE_AUTHENTICATOR_APP_PACKAGE_NAME;
+import static com.microsoft.identity.common.adal.internal.AuthenticationConstants.Broker.BROKER_HOST_APP_PACKAGE_NAME;
+import static com.microsoft.identity.common.adal.internal.AuthenticationConstants.Broker.COMPANY_PORTAL_APP_PACKAGE_NAME;
+
+/**
+ * Key Encryption Manager for Broker.
+ */
+public class AndroidBrokerStorageEncryptionManager extends StorageEncryptionManager {
+    private static final String TAG = AndroidBrokerStorageEncryptionManager.class.getSimpleName();
+
+    /**
+     * A {@link SharedPreferences} key for storing decryption failure event.
+     * see {@link AndroidBrokerStorageEncryptionManager#handleDecryptionFailure}
+     */
+    private static final String CURRENT_ACTIVE_BROKER_SHAREDPREF_KEY = "current_active_broker";
+
+    protected final String LEGACY_AUTHENTICATOR_APP_KEY = "LEGACY_AUTHENTICATOR_APP_KEY";
+    protected final String LEGACY_COMPANY_PORTAL_KEY = "LEGACY_COMPANY_PORTAL_KEY";
+
+    private final Context mContext;
+    private final ITelemetryCallback mTelemetryCallback;
+    private final UserDefinedKeyLoader mLegacyAuthAppKeyLoader;
+    private final UserDefinedKeyLoader mLegacyCPKeyLoader;
+    private final AndroidWrappedKeyLoader mKeyStoreKeyLoader;
+
+    public AndroidBrokerStorageEncryptionManager(@NonNull final Context context,
+                                                 @Nullable final ITelemetryCallback telemetryCallback) {
+        mContext = context;
+        mTelemetryCallback = telemetryCallback;
+
+        mLegacyAuthAppKeyLoader = new UserDefinedKeyLoader(LEGACY_AUTHENTICATOR_APP_KEY,
+                AuthenticationSettings.INSTANCE.getBrokerSecretKeys().get(AZURE_AUTHENTICATOR_APP_PACKAGE_NAME));
+
+        mLegacyCPKeyLoader = new UserDefinedKeyLoader(LEGACY_COMPANY_PORTAL_KEY,
+                AuthenticationSettings.INSTANCE.getBrokerSecretKeys().get(COMPANY_PORTAL_APP_PACKAGE_NAME));
+
+        mKeyStoreKeyLoader = new AndroidWrappedKeyLoader(context, telemetryCallback);
+    }
+
+    // Exposed for Robolectric.
+    protected String getPackageName(){
+        return mContext.getPackageName();
+    }
+
+    @Override
+    public @NonNull AES256KeyLoader getKeyLoaderForEncryption() {
+        if (StorageHelper.sShouldEncryptWithKeyStoreKey) {
+            return mKeyStoreKeyLoader;
+        }
```

Comment:
I don't like that we're accessing a static from `StorageHelper` in this class

Reply:
this is intentional - I don't want to touch the key migration logic right now.

Reply:
Eventually, i think we'll have to revisit all this when we migrate to a single broker model (and I feel like many of them will be scrapped/vastly changed)

Reply:
If we're directly accessing the static variable, can we insure that writes are visible to multiple reading threads for it?

Reply:
it's a final variable - meant to be a switch to enable/disable this feature (but we never do)